impeller/compiler/compiler.cc

```diff
@@ -210,309 +207,138 @@ static CompilerBackend CreateCompiler(const spirv_cross::ParsedIR& ir,
-    compiler_opts.SetBindingBaseForStage(
-        ToShaderCShaderKind(SourceType::kFragmentShader),
-        static_cast<shaderc_uniform_kind>(uniform_kind), kFragBindingBase);
-  }
-}
-
-Compiler::Compiler(const fml::Mapping& source_mapping,
+Compiler::Compiler(std::shared_ptr<const fml::Mapping> source_mapping,
                    const SourceOptions& source_options,
                    Reflector::Options reflector_options)
     : options_(source_options) {
-  if (source_mapping.GetMapping() == nullptr) {
-    COMPILER_ERROR
+  if (!source_mapping || source_mapping->GetMapping() == nullptr) {
+    COMPILER_ERROR(error_stream_)
         << "Could not read shader source or shader source was empty.";
     return;
   }
 
   if (source_options.target_platform == TargetPlatform::kUnknown) {
-    COMPILER_ERROR << "Target platform not specified.";
-    return;
-  }
-
-  auto shader_kind = ToShaderCShaderKind(source_options.type);
-
-  if (shader_kind == shaderc_shader_kind::shaderc_glsl_infer_from_source) {
-    COMPILER_ERROR << "Could not figure out shader stage.";
+    COMPILER_ERROR(error_stream_) << "Target platform not specified.";
     return;
   }
 
-  shaderc::CompileOptions spirv_options;
+  SPIRVCompilerOptions spirv_options;
 
   // Make sure reflection is as effective as possible. The generated shaders
   // will be processed later by backend specific compilers.
-  spirv_options.SetGenerateDebugInfo();
+  spirv_options.generate_debug_info = true;
 
   switch (options_.source_language) {
     case SourceLanguage::kGLSL:
       // Expects GLSL 4.60 (Core Profile).
       // https://www.khronos.org/registry/OpenGL/specs/gl/GLSLangSpec.4.60.pdf
-      spirv_options.SetSourceLanguage(
-          shaderc_source_language::shaderc_source_language_glsl);
-      spirv_options.SetForcedVersionProfile(
-          460, shaderc_profile::shaderc_profile_core);
+      spirv_options.source_langauge =
+          shaderc_source_language::shaderc_source_language_glsl;
+      spirv_options.source_profile = SPIRVCompilerSourceProfile{
+          shaderc_profile::shaderc_profile_core,  //
+          460,                                    //
+      };
       break;
     case SourceLanguage::kHLSL:
-      spirv_options.SetSourceLanguage(
-          shaderc_source_language::shaderc_source_language_hlsl);
+      spirv_options.source_langauge =
+          shaderc_source_language::shaderc_source_language_hlsl;
       break;
     case SourceLanguage::kUnknown:
-      COMPILER_ERROR << "Source language invalid.";
+      COMPILER_ERROR(error_stream_) << "Source language invalid.";
       return;
   }
 
-  SetLimitations(spirv_options);
-
   switch (source_options.target_platform) {
     case TargetPlatform::kMetalDesktop:
-    case TargetPlatform::kMetalIOS:
-      spirv_options.SetOptimizationLevel(
-          shaderc_optimization_level::shaderc_optimization_level_performance);
+    case TargetPlatform::kMetalIOS: {
+      SPIRVCompilerTargetEnv target;
```

Comment:
It looks like every switch case declares one of these. If you hoist out the declaration you can avoid the `{}` in the cases.

Reply:
May I get this in another patch? The compiler needs some TLC. I'm going to get rid of the unknown platforms with optionals and rework this to explicitly reference Stage1 and Stage2.

Reply:
Yeah. Still lgtm.

Reply:
Does "Stage 1" mean source->spirv + reflection, and "Stage 2" mean spirv->target (ideally in Vulkan's case, spirv->optimized spirv)?

Reply:
Yeah, just the diagram in the readme.